API enable deprecation notices to be disabled

tractorcow opened the pull request:
This should reduce the impact of deprecating non-breaking api early.

tractorcow commented:
Probably should collect feedback before we merge this. @silverstripe/core-team , @hafriedlander any thoughts?

sminnee commented:
Thanks, Damian.

 * Have we posted this to silverstripe-dev?
 * Do we want to add support for an _ss_environment define?
 * There's a minor point that should be acknowledged: in removing deprecation notices, we changing behaviour in a minor release. However, because this change actively reduces upgrade friction, which is the goal of semver, I think this is okay.

tractorcow commented:
@sminnee done :D

tractorcow commented:
Problem I've found in this... Config class uses Deprecation, so we can't use any `Config` api in here. I'll have to strip it out. =(

tractorcow commented:
Here is one benefit of this:

![image](https://cloud.githubusercontent.com/assets/936064/8074401/0f9ed870-0f86-11e5-8dea-c92db571bc2a.png)

tractorcow commented:
Finally fixing issues with deprecation in some tests. Hope I have this finally....

gregsmirnov commented:
@tractorcow small improvement. You missed method reference in dockblock for dumping settings

::dump_version_settings => ::dump_settings
::restore_version_settings => ::restore_settings

willmorgan commented:
This is a good PR, although I think deprecation notices should be shown by default in dev mode. Agree that they should never be shown in live mode.

dhensby commented on dev/Deprecation.php:
```diff
@@ -116,6 +123,34 @@ protected static function get_called_method_from_trace($backtrace, $level = 1) {
+	 */
+	public static function get_enabled() {
+		// Never raise deprecation notices in a live environment
+		if(Director::isLive(true)) {
```
Personally my preference here would be `!Director::isDev()`

I don't see that deprecation notices should come up under any other circumstance

oddnoc replied:
Director::isLive() takes no arguments.

tractorcow replied:
Ah, it looks like it used to; I'll update. ;D

tractorcow replied:
@dhensby agreed, let's make it !isDev(). Test sites shouldn't show deprecation notices.

dhensby commented on docs/en/04_Changelogs/3.2.0.md:
```diff
@@ -19,6 +19,7 @@
  * `FormField::validate` now requires an instance of `Validator`
  * Implementation of new "Archive" concept for page removal, which supercedes "delete". Where deletion removed
    pages only from draft, archiving removes from both draft and live simultaneously.
+ * `Deprecation` class no longer raises issues automatically. Enable `Deprecation.enabled` config to turn on.
```
This patch doesn't appear to use the `Config` system

dhensby commented:
@willmorgan As in, if dev, show unless explicitly off. If not dev hide unless explicitly shown?

Except really we are saying they should never be shown outside dev environments.

While I agree that deprecations should be on by default, consensus does not lie with that point of view. Most frameworks will only show these kinds of notice is "debug" mode, so that's essentially what we have here.

willmorgan commented:
> As in, if dev, show unless explicitly off. If not dev hide unless explicitly shown?

No - if dev, show unless explicitly off. If not dev, never show.

tractorcow commented:
I think that popular opinion is to favour on by default, but only in dev. I think that we should leave it on by default (as it is) and perhaps look at changing the default setting in a later framework version?

tractorcow commented:
I've updated this based on feedback:

- Fixed docs to not refer to config anymore.
- Fixed phpdoc
- Deprecation only available at all on dev
- Deprecation still enabled by default

dhensby commented:
I think the two big issues here are: 

1. Depreciation errors break tests (believe this is now fixed in this PR?) 
2. The SS error handler shouts too loudly about them  

I think the ss error handler should be much more subtle about them, perhaps suggesting how they are disabled and what it means and staying as the default php line of text. 

Further, a depreciation notice should probably only be fired once per function, not repeatedly during execution. (this is WP behaviour)

dhensby commented:
A +1 for this too is that it is BC! If we want them off by default, then it should happen in v4

dhensby commented:
So, what does this mean for `Deprecation::notice($version)`, should $version say when we expect to take it out or when it was deprecated? The former means devs will have to actively change their notice settings to ever be warned, the latter doesn't

jonom commented:
Shouldn't `$version` indicate when the function is expected to be removed? The notice should give you an idea of how urgent it is to change I think. Agree a much more subtle notice would be good too, particularly in the case of a module causing a deprecation notice you can't do anything about, especially if the module owner doesn't want to create multiple branches just to get rid of deprecation notices.

On that subject could something like this be an idea to give us a bit more flexibility about which notices to let through?
```php
/**
 * Specify classes which should have deprecation notices muted.
 * Deprecation is only available on dev.
 *
 * @var array
 */
protected static $suppressed = array();
```
```yml
Deprecation:
  suppressed:
    - SomeOldClass
```

dhensby commented on dev/Deprecation.php:
```diff
@@ -116,20 +126,47 @@ protected static function get_called_method_from_trace($backtrace, $level = 1) {
+		if(isset(self::$enabled)) {
+			return self::$enabled;
+		}
+		if(defined('SS_DEPRECATION_ENABLED')) {
```
Should the static var override the constant? I generally the constants take absolute precedence usually.

However, I'd suggest they have equal weight. if either is disabled, then don't show? Maybe that makes it a bit less flexible, though...

tractorcow replied:
Constants are "environment" and static vars "projects", which is the correct order of inheritance.

hafriedlander commented:
`$version` is / was supposed to be when it's expected to be taken out. The commit the `Deprecation::notice` call was added is when the deprecation itself occurs. Was more important before semver (now the rule is presumably "next major version").

tractorcow commented:
> Depreciation errors break tests (believe this is now fixed in this PR?)

If we merge this, we can disable them in travis-support, but developers will still get them on by default.

hafriedlander commented on dev/Deprecation.php:
```diff
@@ -43,6 +43,16 @@ class Deprecation {
+	 * Override whether deprecation is enabled. If null, then fallback to 
+	 * SS_DEPRECATION_ENABLED, and then true if not defined.
+	 * 
+	 * Deprecation is only available on dev.
```
Should probably add a note here that we can't use Config system because of infinite loop risk.

tractorcow replied:
Added in updated push.

tractorcow commented:
Updated @hafriedlander

kinglozzer commented:
Sorry I’m a few days late to the deprecation party...

+1 for on by default in dev, off in everything else (I think there’s already a consensus on that anyway...)

>So, what does this mean for Deprecation::notice($version), should $version say when we expect to take it out or when it was deprecated? The former means devs will have to actively change their notice settings to ever be warned, the latter doesn't

Do we definitely have an agreement on this? I’m confused about who is voting for which approach.

I’ve changed my mind and am now advocating the *“`$version` = when it was first deprecated”* approach, so that devs *don’t* have to manually change their notice version to see the warnings. The only downside I can see is that if I drop in a 3.2 update to a 3.1 site, I may start seeing deprecation notices popping up. However, that would really be a non-issue if this PR is merged as they’d be easy to turn off.

tractorcow commented:
I think this is a bit late for 3.2; It's ok to look at this in 3.3, or otherwise just 4.

dhensby commented:
We should be setting the version to when it was deprecated, not when it will be taken out. There's no obligation to take deprecated functions out at any specific release (though it can only be major ones)

dhensby commented:
I don't think this is too late for 3.2. 

Its an issue that's been bugging the community for a while and if it waits to 3.3, people could be waiting until next year for this to make it into a stable release

tractorcow commented:
Closing and re-opening on 3.2 branch.

tractorcow commented:
See https://github.com/silverstripe/silverstripe-framework/pull/4403
